turbonomic/kubeturbo
# init impl of static pod daemon sets

bug: https://vmturbo.atlassian.net/browse/OM-90831

# References:
https://kubernetes.io/docs/tasks/configure-pod-container/static-pod/
https://cloud.google.com/kubernetes-engine/docs/concepts/network-overview

# background:
 We discovered that rather than implementing kube-proxy with daemon sets, GKE has used static pods on every node. This change will either check every node or every node pool for pods with the same intent. 
 
 
# description:
 Before dispatching workers the code will find all the static pods acting as daemons and mark them as such. We will find the different groups of pods by a common prefix that the user would have to consistently use during creation on the node. After all are found the code keeps a mapping in memory of the static pods that are also daemons for creation during pod building.

# Testing:
- build the docker image
- deploy to your docker hub account
- deploy to any GKE cluster
- move all pods except the mirror pod to other nodes
    - Use kubectl cordon or kubectl drain to remove all pods except for the mirror pod (You wouldn't be able to move the mirror pod).
 - notice that the VM will have a suspend action for both the pod and the VM 
 
 
 Afterwords the UI should show something like below:
 
<img width="2471" alt="Screen Shot 2022-10-18 at 9 50 58 AM" src="https://user-images.githubusercontent.com/8473487/196448989-161a5f0f-6ca8-459a-b4b1-4e00fc829fa8.png">
